src/node/timeoffsets.h

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) 2024-present The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_NODE_TIMEOFFSETS_H
+#define BITCOIN_NODE_TIMEOFFSETS_H
+
+#include <sync.h>
+
+#include <atomic>
+#include <chrono>
+#include <deque>
+#include <stddef.h>
+
+class TimeOffsets
+{
+private:
+    static constexpr size_t N{50};
+    //! Minimum difference between system and network time for a warning to be raised.
+    static constexpr std::chrono::minutes m_warn_threshold{10};
+    mutable std::atomic<bool> m_warning_emitted{false};  // Avoid flooding the user with warnings.
```

Comment:
For GUI and RPC warnings, this PR pretty much preserves the current behaviour: GUI messagebox gets shown only once (until restarted), and the RPC will keep firing forever (on master, it will be reset when another `SetMiscWarning()` is called, but in practice I don't think that really happens frequently).

For the debug log, however, this PR quite significantly improves things by
1) doing an actual rolling calculation, which means that if clocks start drifting, the user will eventually still get a warning, whereas on master the time adjustment system completely stops after the first 199 outbound handshakes
2) producing repeated log warnings (one for every outbound handshake) for as long _and only for as long_ as the median time offset exceeds the threshold.

While I agree that for GUI & RPC things should be improved, this PR is no regression, and I'd rather it be done in a separate PR. Avoiding flooding seems less trivial to me than adding a simple timer: having a GUI message popup every 10 minutes is quite annoying if someone decides to flood the network with nodes with wrong clocks.

> Is there a way for the user to see the median network time, as perceived by his node?
yes, through `getnetworkinfo` a user can see the median offset

> 1 month later the issues resurface - there is no warning or indication about it?
there will be warning logs every time we handshake with an outbound peer and the median timeoffset exceeds the threshold

Reply:
See https://github.com/stickies-v/bitcoin/commit/0146553aaab53b6092bb54e0b2c71282fef44787 for a potential implementation on turning {GUI, RPC} warnings off again when the offsets fall within the threshold again. Happy to open a follow-up PR with this if you like the approach, I think this PR is too close to merge to overhaul it again as per my previous comment, what do you think?